05_debug/deb_module.c

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/printk.h>
+#include <linux/moduleparam.h>
+#include <linux/kernel.h>
+#include <linux/sysfs.h>
+#include <linux/proc_fs.h>
+#include <linux/types.h>
+#include <linux/uaccess.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include "/home/ivan/repos/u-boot/include/linux/ctype.h"
+#include <linux/vmalloc.h>
+
+MODULE_AUTHOR("Ivan Kryvosheia");
+MODULE_DESCRIPTION("Debug module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+MODULE_VERSION("1.0");
+
+#define SYSFS_DIR_NAME		"sysfs_gl"
+#define PROCFS_DIR_NAME		"procfs_gl"
+
+static ssize_t sysfs_show(struct kobject *kobj,
+		struct kobj_attribute *attr, char *buf);
+static ssize_t sysfs_store(struct kobject *kobj,
+		struct kobj_attribute *attr, const char *buf, size_t count);
+
+static ssize_t	proc_read(struct file *file, char __user *buf,
+		size_t count, loff_t *ppos);
+static ssize_t	proc_write(struct file *file, const char __user *ubuf,
+		size_t count, loff_t *ppos);
+
+static int init_sysfs_attr(void);
+static void deinit_sysfs_attr(void);
+
+static int init_procfs_attr(void);
+static void deinit_procfs_attr(void);
+
+static void free_memory(void);
+
+const static struct kobj_attribute sysfs_attr = {
+	.attr = { .name = "conv", .mode = 0666 },
+	.show	= sysfs_show,
+	.store	= sysfs_store
+};
+
+const static struct file_operations procfs_attr = {
+	.owner = THIS_MODULE,
+	.read = proc_read,
+	.write = proc_write,
+};
+
+struct my_obj {
+	char *str;
+} sysfs_obj, procfs_obj;
+
+//sysfs
+struct kobject *kobj_inst;
+
+//procfs
+struct proc_dir_entry *proc_dir;
+
+static int init_procfs_attr(void)
+{
+	/* create directory in /proc/ */
+	proc_dir = proc_create(PROCFS_DIR_NAME, 0660, NULL, &procfs_attr);
+	if (!proc_dir)
+		return -ENOMEM;
+
+	return 0;
+}
+
+static void deinit_procfs_attr(void)
+{
+	proc_remove(proc_dir);
+}
+
+ssize_t	proc_read(struct file *file, char __user *buf, size_t count,
+		loff_t *ppos)
+{
+	if (procfs_obj.str)
+		pr_info("%s\n", procfs_obj.str);
+
+	return 0;
+}
+
+ssize_t	proc_write(struct file *file, const char __user *ubuf,
+		size_t count, loff_t *ppos)
+{
+	int i = 0;
+
+	kfree(procfs_obj.str);
+
+	procfs_obj.str = kmalloc((count + 1) * sizeof(char), GFP_KERNEL);
```

Comment:
consider using kstrdup(buf, GFP_KERNEL);
it will make life easier

Reply:
![Selection_015](https://user-images.githubusercontent.com/23324355/112123034-8fab7d00-8bc9-11eb-9136-11256003b6df.png)

Reply:
I dont know whats "behind the hood" but one needs to "use copy_from_user()" function to copy data from callback.
I have broken my qemu during trying to do the opposite,while I havent seen example from the internet I could not get rid of these exceptions.
The print, I've left here is current one while, I replaced kmalloc and copy_from_user() with kstrdup(buf, GFP_KERNEL).
So i am going to left as it is in the code.

Reply:
Yes, you're right! This is procfs handler, so it requires copying from user.
Indeed kstrdup() will not work here, possible to use strndup_user() function.
But let's just leave it as is, copy_from_user() is good enough here.